PythonVisualizations/QuadTree.py

```diff
@@ -0,0 +1,227 @@
+    def drawLine(self, n, parent, direction):
+        if not self.direction: 
+            n.horizontal_line =2, n.y, 801, n.y
+            n.vertical_line = n.x, 2, n.x, 401
+        
+        else:
+            p_Hx0, p_Hy0, p_Hx1, p_Hy1 = parent.horizontal_line
+            p_Vx0, p_Vy0, p_Vx1, p_Vy1 = parent.vertical_line
+            S_vertical_line = n.x, parent.y, n.x, p_Vy1 
+            N_vertical_line = n.x, p_Vy0, n.x, parent.y
+            W_horizontal_line = p_Hx0, n.y, parent.x, n.y
+            E_horizontal_line = parent.x, n.y, p_Hx1 , n.y           
+            if self.direction == NE: 
+                n.horizontal_line = E_horizontal_line 
+                n.vertical_line = N_vertical_line
+            elif self.direction== NW: 
+                n.horizontal_line = W_horizontal_line 
+                n.vertical_line = N_vertical_line
+            elif self.direction == SW: 
+                n.horizontal_line = W_horizontal_line 
+                n.vertical_line = S_vertical_line
+            elif self.direction == SE: 
+                n.horizontal_line = E_horizontal_line
+                n.vertical_line = S_vertical_line
+            self.direction = None
+        if self.showTree == False: return
+        return self.canvas.create_line(n.horizontal_line, fill = self.LINE_COLOR), self.canvas.create_line(n.vertical_line, fill = self.LINE_COLOR)
+    
+    #Switches the showTree attribute
+    #deletes all the lines and root highlight if showTree is off
+    #Otherwise draws all the lines at appropriate coordinates and the root highlight
+    def graphLineDisplay(self):
+        if self.showTree == True:
+            self.showTree= False
+            for i in self.lines: self.canvas.delete(i)
+            self.lines = []
+            self.canvas.delete(self.rootOutline)
+            
+        else:
+            self.showTree= True
+            for i in self.nodes:
+                if i == self.__root:
+                    self.rootOutline = self.canvas.create_oval((i.x - self.CIRCLE_DIMEN//2-2, i.y- self.CIRCLE_DIMEN//2-2, 
+                                             i.x + self.CIRCLE_DIMEN//2+2, i.y + self.CIRCLE_DIMEN//2 +2), 
+                                            outline = "yellow", width = 3)
+                    self.canvas.lift(self.rootOutline)
+                horiz = self.canvas.create_line(i.horizontal_line, fill = self.LINE_COLOR)
+                vert = self.canvas.create_line(i.vertical_line, fill = self.LINE_COLOR)
+                self.canvas.lower(horiz), self.canvas.lower(vert)
+                self.lines.append(horiz), self.lines.append(vert)
+        
+    
+    #wrapper method for insert   
+    def insert(self, x, y, d): 
+        self.__root = self.__insert(self.__root, x, y, d)
+    
+    #creates a new node if none at designated coords
+    #otherwise adds data to existing coordinate
+    def __insert(self, n, x, y, d):
+        callEnviron = self.createCallEnvironment()
```

Comment:
I think the insert routine is the right place to create a unique data label, `d`.  It should be done with a loop that exits if `d` is unique and increments the `COUNTER` and sets `d = 'P' + str(COUNTER)` in the loop body.  This will fix the "Naming bug".

The `clickInsert()` method is the right place to catch attempts to insert nodes with duplicate labels, as you have already done.